core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedBeanFactoryRegistry.java

```diff
@@ -115,13 +118,28 @@ static AnnotatedBeanFactory<?> find(@Nullable BeanFactoryId beanFactoryId) {
 
     static Set<AnnotatedValueResolver> uniqueResolverSet() {
         return new TreeSet<>((o1, o2) -> {
-            final String o1Name = o1.httpElementName();
-            final String o2Name = o2.httpElementName();
-            if (o1Name.equals(o2Name) && o1.annotationType() == o2.annotationType()) {
+            String o1Name = o1.httpElementName();
+            String o2Name = o2.httpElementName();
+            final Class<? extends Annotation> o1AnnotationType = o1.annotationType();
+            final Class<? extends Annotation> o2AnnotationType = o2.annotationType();
+            if (o1Name.equals(o2Name) && o1AnnotationType == o2AnnotationType) {
                 return 0;
             }
-            // We are not ordering, but just finding duplicate elements.
-            return -1;
+
+            // TreeSet internally creates a binary tree. A consistent comparison for the two inputs is
+            // necessary to traverse the binary tree correctly and check uniqueness.
+            // If compare(o1, o2) returns -1, compare(o2, o1) should return 1 or a positive number.
+            if (o1AnnotationType != null) {
+                o1Name += '/' + o1AnnotationType.getName();
+            }
+            if (o2AnnotationType != null) {
+                o2Name += '/' + o2AnnotationType.getName();
+            }
+            final int result = o1Name.compareTo(o2Name);
+            if (result != 0) {
+                return result;
+            }
+            return Ints.compare(Objects.hashCode(o1AnnotationType), Objects.hashCode(o2AnnotationType));
```

Comment:
Question: can't we just return `result` instead of calling this?

Reply:
In rare cases, a call path reaches here. If the two different class loaders may load a class, the reference is different for the two classes but `o1Name` and `o2Name` are equal. I wanted to avoid a situation where `o1Name.equals(o2Name) && o1AnnotationType == o2AnnotationType` is false and then returns 0 here.

Reply:
Do you mean when two different class loaders load the same class?

Reply:
Right, but it would not be a common case.

Reply:
Had a chat with @ikhoon and I understand the situation now.